src/openfermion/utils/_operator_utils.py

```diff
@@ -846,3 +848,97 @@ def normal_ordered(operator, hbar=1.):
+            compatible sub-operator exists). A sub-operator is compatible
+            if there exists a tensor product basis in which the term,
+            and all terms already in the sub-operator, are diagonal.
+        seed (int): default None. If None, the random number generator
+            is not re-initialised. Otherwise, the random number generator
+            is re-initialised with the seed.
+
+    Returns:
+        sub_operators (dict): a dictionary where each key defines a
+            tensor product basis, and each corresponding value is a
+            QubitOperator with terms that are all diagonal in
+            that basis.
+            **key** (tuple of tuples): Each key is a term, which defines
+                a tensor product basis. A term is a product of individual
+                factors; each factor is represented by a tuple of the form
+                (`index`, `action`), and these tuples are collected into a
+                larger tuple which represents the term as the product of
+                its factors. `action` is from the set {'X', 'Y', 'Z'} and
+                `index` is a non-negative integer corresponding to the
+                index of a qubit.
+            **value** (QubitOperator): A QubitOperator with terms that are
+                diagonal in the basis defined by the key it is stored in.
+
+    Raises:
+       TypeError: Operator of invalid type.
+
+    """
+
+    if not isinstance(operator, QubitOperator):
+        raise TypeError('Can only split QubitOperator into tensor product'
+                        ' basis sets. {} is not supported.'.format(
+                            type(operator).__name__))
+
+    sub_operators = OrderedDict()
+    for term, coefficient in operator.terms.items():
+        inserted = False
+        bases = sub_operators.keys()
+        if randomize:
+            bases = list(bases)
+            if seed:
+                random.seed(seed)
+            random.shuffle(bases)
+
+        for basis in bases:
+            basis_qubits = tuple(zip(*basis))[0] if len(basis) > 0 else []
+            conflicts = tuple((i, P) for (i, P) in term if i in
+                              basis_qubits and (i, P) not in basis)
+            if conflicts:
+                continue
+
+            additions = tuple((i, P) for (i, P) in term if i not in
+                              basis_qubits)
+            if additions:
+                new_basis = tuple(sorted(basis + additions,
+                                         key=lambda factor: factor[0]))
+                sub_operators[new_basis] = sub_operators.pop(basis)
+                sub_operators[new_basis] += QubitOperator(term, coefficient)
+            else:
+                sub_operators[basis] += QubitOperator(term, coefficient)
+
+            inserted = True
+            break
+
+        if not inserted:
+            sub_operators[term] = QubitOperator(term, coefficient)
```

Comment:
Instead of using an inserted=True/False flag to handle the different cases, define a helper function
```
_find_compatible_basis(term, bases)
```
which looks through `bases` for a basis that is compatible with `term` and returns it if it exists, and otherwise returns None.

Reply:
Ok sure good point, I've refactored it.